File: patcha/collectors/window.py
import json
import subprocess
from datetime import datetime, timezone
from pathlib import Path
from typing import List, Tuple

from patcha.config import config
from patcha.db.models import Event, EventType
from patcha.utils.jsonl import trim_jsonl

_MAX_LOG_ROWS = 100_000
_TRIM_EVERY = 1_000


class WindowCollector:
    """Tracks active macOS app and window title via osascript.

    requires accessibility permission:
    System Settings > Privacy & Security > Accessibility → enable terminal/patcha.
    """

    _SCRIPT = Path(__file__).parent.parent / "macos" / "window_title.applescript"

    def __init__(self):
        self.log_file: Path = config.data_dir / "window_log.jsonl"
        self._write_count: int = 0

    def _get_active_window(self) -> Tuple[str, str]:
        try:
            result = subprocess.run(
                ["osascript", self._SCRIPT],
                capture_output=True,
                text=True,
                timeout=5,
            )
            if result.returncode == 0:
                parts = result.stdout.strip().split("|||")
                app = parts[0].strip()
                title = parts[1].strip() if len(parts) > 1 else ""
                return app, title
        except (subprocess.TimeoutExpired, OSError):
            pass
        return "", ""

    def record_current_window(self) -> None:
        """Snapshot the active window and append to the log file."""
        app, title = self._get_active_window()
        if not app:
            return

        config.data_dir.mkdir(parents=True, exist_ok=True)
        entry = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "app": app,
            "title": title,
        }
        with open(self.log_file, "a") as f:
            f.write(json.dumps(entry) + "\n")

        self._write_count += 1
        if self._write_count % _TRIM_EVERY == 0:
            trim_jsonl(self.log_file, _MAX_LOG_ROWS)

    def collect_windows(self, since: datetime) -> List[Event]:
        """Return completed window-focus sessions logged since `since`.

        A session ends when the active app or window title changes. The
        currently active (still-open) session is excluded since its
        duration is not yet known.
        """
        if not self.log_file.exists():
            return []

        entries: List[Tuple[datetime, str, str]] = []
        with open(self.log_file) as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    obj = json.loads(line)
                    ts = datetime.fromisoformat(obj["timestamp"])
                    if ts >= since:
                        entries.append((ts, obj["app"], obj.get("title", "")))
                except (json.JSONDecodeError, KeyError, ValueError):
                    continue

        if len(entries) < 2:
            return []

        events: List[Event] = []
        session_start, session_app, session_title = entries[0]

        for ts, app, title in entries[1:]:
            if app != session_app or title != session_title:
                duration = int((ts - session_start).total_seconds())
                if duration >= 30:
                    events.append(
                        self._make_event(
                            session_start, session_app, session_title, duration
                        )
                    )
                session_start, session_app, session_title = ts, app, title

        return events

    def _make_event(
        self, timestamp: datetime, app: str, title: str, duration_seconds: int
    ) -> Event:
        label = title if title else app
        raw = f"{app}: {label} ({duration_seconds}s)"
        return Event(
            timestamp=timestamp,
            type=EventType.WINDOW,
            source="window",
            raw_content=raw,
            metadata={
                "app_name": app,
                "window_title": title,
                "duration_seconds": duration_seconds,
            },
        )

File: rust/patcha/src/collectors/window.rs
use crate::models::{Event, EventType};
use anyhow::Result;
use chrono::{DateTime, Utc};
use std::io::Write;
use std::path::PathBuf;

const MAX_LOG_ROWS: usize = 100_000;
const TRIM_EVERY: usize = 1_000;
const MIN_DURATION_SECS: i64 = 30;

pub struct WindowCollector {
    log_file: PathBuf,
    script_path: PathBuf,
    write_count: usize,
}

impl WindowCollector {
    pub fn new(data_dir: &std::path::Path, script_path: PathBuf) -> Self {
        Self {
            log_file: data_dir.join("window_log.jsonl"),
            script_path,
            write_count: 0,
        }
    }

    /// Snapshot the active window and append to the log file.
    /// Called periodically by the daemon (same cadence as ax_content).
    pub fn record_current_window(&mut self) -> Result<()> {
        let (app, title) = self.get_active_window()?;
        if app.is_empty() {
            return Ok(());
        }

        if let Some(parent) = self.log_file.parent() {
            std::fs::create_dir_all(parent)?;
        }

        let entry = serde_json::json!({
            "timestamp": Utc::now().to_rfc3339(),
            "app": app,
            "title": title,
        });

        let mut f = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.log_file)?;
        writeln!(f, "{}", entry)?;

        self.write_count += 1;
        if self.write_count.is_multiple_of(TRIM_EVERY) {
            trim_jsonl(&self.log_file, MAX_LOG_ROWS)?;
        }

        Ok(())
    }

    /// Return completed window-focus sessions since `since`.
    /// The currently active (still-open) session is excluded.
    pub fn collect_windows(&self, since: DateTime<Utc>) -> Vec<Event> {
        let Ok(entries) = self.read_log(since) else {
            return Vec::new();
        };
        if entries.len() < 2 {
            return Vec::new();
        }

        let mut events = Vec::new();
        let (mut session_start, mut session_app, mut session_title) = entries[0].clone();

        for (ts, app, title) in &entries[1..] {
            if app != &session_app || title != &session_title {
                let duration = (*ts - session_start).num_seconds();
                if duration >= MIN_DURATION_SECS {
                    events.push(make_window_event(
                        session_start,
                        &session_app,
                        &session_title,
                        duration,
                    ));
                }
                session_start = *ts;
                session_app = app.clone();
                session_title = title.clone();
            }
        }
        // The last session (still active) is intentionally not emitted.

        events
    }

    // -----------------------------------------------------------------------
    // Helpers
    // -----------------------------------------------------------------------

    fn get_active_window(&self) -> Result<(String, String)> {
        if !self.script_path.exists() {
            return Ok((String::new(), String::new()));
        }

        let output = std::process::Command::new("osascript")
            .arg(&self.script_path)
            .output();

        match output {
            Ok(out) if out.status.success() => {
                let text = String::from_utf8_lossy(&out.stdout);
                let parts: Vec<&str> = text.trim().split("|||").collect();
                let app = parts.first().copied().unwrap_or("").trim().to_owned();
                let title = parts.get(1).copied().unwrap_or("").trim().to_owned();
                Ok((app, title))
            }
            _ => Ok((String::new(), String::new())),
        }
    }

    fn read_log(&self, since: DateTime<Utc>) -> Result<Vec<(DateTime<Utc>, String, String)>> {
        if !self.log_file.exists() {
            return Ok(Vec::new());
        }

        let content = std::fs::read_to_string(&self.log_file)?;
        let mut entries = Vec::new();

        for line in content.lines() {
            let line = line.trim();
            if line.is_empty() {
                continue;
            }
            let Ok(obj) = serde_json::from_str::<serde_json::Value>(line) else {
                continue;
            };
            let Some(ts_str) = obj.get("timestamp").and_then(|v| v.as_str()) else {
                continue;
            };
            let Ok(ts) = DateTime::parse_from_rfc3339(ts_str) else {
                continue;
            };
            let ts = ts.with_timezone(&Utc);
            if ts < since {
                continue;
            }
            let app = obj
                .get("app")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_owned();
            let title = obj
                .get("title")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_owned();
            entries.push((ts, app, title));
        }

        Ok(entries)
    }
}

fn make_window_event(
    timestamp: DateTime<Utc>,
    app: &str,
    title: &str,
    duration_seconds: i64,
) -> Event {
    let label = if title.is_empty() { app } else { title };
    let raw = format!("{app}: {label} ({duration_seconds}s)");

    let mut e = Event::new(EventType::Window, raw);
    e.timestamp = timestamp;
    e.source = Some("window".to_owned());
    e.metadata.insert("app_name".into(), serde_json::json!(app));
    e.metadata
        .insert("window_title".into(), serde_json::json!(title));
    e.metadata.insert(
        "duration_seconds".into(),
        serde_json::json!(duration_seconds),
    );
    e
}

fn trim_jsonl(path: &std::path::Path, max_rows: usize) -> Result<()> {
    let content = std::fs::read_to_string(path)?;
    let lines: Vec<&str> = content.lines().collect();
    if lines.len() <= max_rows {
        return Ok(());
    }
    let keep = &lines[lines.len() - max_rows..];
    let tmp = path.with_extension("jsonl.tmp");
    std::fs::write(&tmp, keep.join("\n") + "\n")?;
    std::fs::rename(tmp, path)?;
    Ok(())
}
